src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs

```diff
@@ -460,6 +482,58 @@ public void ExitMSBuildCallbackState()
             _inMSBuildCallback = false;
         }
 
+        /// <summary>
+        /// Requests CPU resources from the scheduler.
+        /// </summary>
+        public int RequestCores(object monitorLockObject, int requestedCores, bool waitForCores)
+        {
+            ErrorUtilities.VerifyThrow(Monitor.IsEntered(monitorLockObject), "Not running under the given lock");
+            VerifyIsNotZombie();
+
+            ResourceResponse responseObject = null;
+            using AutoResetEvent responseEvent = new AutoResetEvent(false);
+            _pendingResourceRequests.Enqueue((ResourceResponse response) =>
```

Comment:
How can you get multiple requests in this queue? The first one is non-blocking (so  the queue gets one item which is removed before this method exists), and subsequent calls are blocking. May be worth making your answer a code comment :)